### PR TITLE
Introduce eofparse and eofparsefuzz

### PR DESCRIPTION
**Pull request by axic**
Pulled this out from #563. Could be merged on its own.

### CODE REVIEW

**Review by rodiazet — APPROVED** [review]

File: test/eofparse/eofparse.cpp [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    {
+        for (std::string line; std::getline(std::cin, line);)
+        {
+            if (line.empty() || line.starts_with('#'))
+                continue;
+
+            auto o = from_hex2(line);
+            if (!o)
+            {
+                std::cout << "err: invalid hex" << std::endl;
```
Comment (gumb0) [inline test/eofparse/eofparse.cpp:55]:
```suggestion
                std::cout << "err: invalid hex\n";
```

File: test/eofparse/eofparse.cpp [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+                continue;
+            }
+
+            const auto eof = std::move(*o);
+            const auto err = evmone::validate_eof(EVMC_CANCUN, eof);
+            if (err != evmone::EOFValidationError::success)
+            {
+                std::cout << "err: " << evmone::get_error_message(err) << std::endl;
```
Comment (gumb0) [inline test/eofparse/eofparse.cpp:63]:
```suggestion
                std::cout << "err: " << evmone::get_error_message(err) << "\n";
```

File: test/eofparse/eofparse.cpp [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+                continue;
+            }
+
+            const auto header = evmone::read_valid_eof1_header(evmone::bytes_view{eof}.begin());
+            std::cout << "OK "
+                      << evmc::hex(evmc::bytes_view{&eof[header.code_begin()], header.code_size})
+                      << std::endl;
```
Comment (gumb0) [inline test/eofparse/eofparse.cpp:70]:
```suggestion
                      << "\n";

File: test/eofparse/eofparse.cpp [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+            const auto eof = std::move(*o);
+            const auto err = evmone::validate_eof(EVMC_CANCUN, eof);
+            if (err != evmone::EOFValidationError::success)
+            {
+                std::cout << "err: " << evmone::get_error_message(err) << std::endl;
+                continue;
+            }
+
+            const auto header = evmone::read_valid_eof1_header(evmone::bytes_view{eof}.begin());
```
Comment (gumb0) [inline test/eofparse/eofparse.cpp:67]:
```suggestion
            const auto header = evmone::read_valid_eof1_header(eof.begin());
```

File: test/eofparse/eofparse.cpp [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+{
+inline constexpr bool isalnum(char ch) noexcept
+{
+    return (ch >= '0' && ch <= '9') || (ch >= 'A' && ch <= 'Z') || (ch >= 'a' && ch <= 'z');
+}
+
+template <typename BaseIterator>
+struct skip_nonalnum_iterator : evmc::filter_iterator<BaseIterator, isalnum>
+{
+    using evmc::filter_iterator<BaseIterator, isalnum>::filter_iterator;
```
Comment (gumb0) [inline test/eofparse/eofparse.cpp:20]: surprised that such thing is in public evmc API

File: test/eofparse/eofparse.cpp [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+};
+
+template <typename BaseIterator>
+skip_nonalnum_iterator(BaseIterator, BaseIterator) -> skip_nonalnum_iterator<BaseIterator>;
+
+template <typename InputIterator>
+std::optional<evmc::bytes> from_hex2(InputIterator begin, InputIterator end) noexcept
```
Comment (gumb0) [inline test/eofparse/eofparse.cpp:27]: maybe
```suggestion
std::optional<evmc::bytes> from_hex_skip_nonalnum(InputIterator begin, InputIterator end) noexcept
```

File: test/eofparse/eofparse.cpp [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+{
+    evmc::bytes bs;
+    if (!from_hex(skip_nonalnum_iterator{begin, end}, skip_nonalnum_iterator{end, end},
+            std::back_inserter(bs)))
+        return {};
+    return bs;
+}
+
+[[maybe_unused]] auto from_hex2(std::string_view s) noexcept
```
Comment (gumb0) [inline test/eofparse/eofparse.cpp:36]: this is used only once, I would just call the main helper directly with iterators

File: test/eofparse/eofparse.cpp (hunk 4 above)
Comment (axic) [reply test/eofparse/eofparse.cpp:67]: Explicit conversion is needed:
```
/Projects/evmone/test/eofparse/eofparse.cpp:67:64: error: no viable conversion from 'std::basic_string<unsigned char>::const_iterator' (aka '__wrap_iter<const unsigned char *>') to 'bytes_view::const_iterator' (aka 'const unsigned char *')
            const auto header = evmone::read_valid_eof1_header(eof.begin());
                                                               ^~~~~~~~~~~
/Projects/evmone/lib/evmone/eof.hpp:31:32: note: passing argument to parameter 'code' here
    bytes_view::const_iterator code) noexcept;
                               ^
1 error generated.
```

File: test/eofparse/eofparse.cpp (hunk 1 above)
Comment (axic) [reply test/eofparse/eofparse.cpp]: Done.

File: test/eofparse/eofparse.cpp [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+            const auto err = evmone::validate_eof(EVMC_CANCUN, eof);
+            if (err != evmone::EOFValidationError::success)
+            {
+                std::cout << "err: " << evmone::get_error_message(err) << std::endl;
+                continue;
+            }
+
+            const auto header = evmone::read_valid_eof1_header(evmone::bytes_view{eof}.begin());
+            std::cout << "OK "
+                      << evmc::hex(evmc::bytes_view{&eof[header.code_begin()], header.code_size})
```
[the remark this answers is not in the document]
Comment (axic) [reply test/eofparse/eofparse.cpp]: Done.

File: test/eofparse/eofparse.cpp (hunk 6 above)
Comment (axic) [reply test/eofparse/eofparse.cpp]: Done.

File: test/eofparse/eofparse.cpp (hunk 7 above)
Comment (axic) [reply test/eofparse/eofparse.cpp]: Done.

**Pull request merged**